source/agora/test/MissingPreImageDetection.d

```diff
@@ -0,0 +1,597 @@
+/*******************************************************************************
+
+    Contains tests for the validators not revealing their pre-images. There
+    are four cases that the validators do not reveal as follows.
+
+    (A) Never send any pre-image after its initial enrollment
+    (B) Only send one pre-image after enrolling then no more
+    (C) Send pre-image for half of its cycle (potentially in one burst)
+        then stop sending pre-image
+    (D) Send pre-image for half its cycle, stop it for a quarter of the cycle,
+        resume until the end
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.MissingPreImageDetection;
+
+import agora.common.crypto.Key;
+import agora.common.Config;
+import agora.common.Hash;
+import agora.consensus.data.Block;
+import agora.consensus.data.Params;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.Transaction;
+import agora.consensus.EnrollmentManager;
+import agora.test.Base;
+
+import core.stdc.time;
+import core.thread;
+import geod24.Registry;
+
+version (unittest):
+
+/*******************************************************************************
+
+    Make transactions using the old transactoins passed in as a parameter
+
+    Params:
+        old_txs = the old transaction with which new transactions will be made
+        address = the address to refund the old transactions
+
+    Returns:
+        The new transactions generated
+
+*******************************************************************************/
+
+private Transaction[] makeTransactions (Transaction[] old_txs, PublicKey address)
+{
+    return old_txs
+        .length.iota.map!(idx => TxBuilder(old_txs[idx], cast(uint)0))
+        .map!(txb => txb.refund(address).sign())
+        .array;
```

Comment:
That's a very convoluted way of writing:
```suggestion
    return old_txs.map!(tx => TxBuilder(tx, 0).refund(address).sign()).array();
```

Although I'm not sure why you only want to spend a single output ?

Reply:
Will update. No special reason for a single output. But, I will update it also after referring to other codes, If needed. Ty.